code/game/machinery/atm_ret_field.dm

```diff
@@ -0,0 +1,246 @@
+	icon = 'icons/obj/atm_fieldgen.dmi'
+	icon_state = "arfg_off"
+	anchored = TRUE
+	opacity = FALSE
+	density = FALSE
+	power_channel = ENVIRON	//so they shut off last
+	use_power = USE_POWER_IDLE
+	idle_power_usage = 10
+	active_power_usage = 2500
+	var/ispowered = TRUE
+	var/isactive = FALSE
+	var/wasactive = FALSE		//controls automatic reboot after power-loss
+	var/alwaysactive = FALSE	//for a special subtype
+
+	//how long it takes us to reboot if we're shut down by an EMP - mild variance makes it look a bit more interesting and unpredictable
+	var/reboot_delay_min = 50
+	var/reboot_delay_max = 75
+
+	var/hatch_open = FALSE
+	var/wires_intact = TRUE
+	var/list/areas_added
+	var/field_type = /obj/structure/atmospheric_retention_field
+	circuit = /obj/item/circuitboard/arf_generator
+
+/obj/machinery/atmospheric_field_generator/impassable
+	desc = "An older model of ARF-G that generates an impassable retention field. Works just as well as the modern variety, but is slightly more energy-efficient.<br><br>Note: prolonged immersion in active atmospheric retention fields may have negative long-term health consequences."
+	active_power_usage = 2000
+	field_type = /obj/structure/atmospheric_retention_field/impassable
+
+/obj/machinery/atmospheric_field_generator/perma
+	name = "static atmospheric retention field generator"
+	desc = "A floor-mounted piece of equipment that generates an atmosphere-retaining energy field when powered and activated. This model is designed to always be active, though the field will still drop from loss of power or electromagnetic interference.<br><br>Note: prolonged immersion in active atmospheric retention fields may have negative long-term health consequences."
+	alwaysactive = TRUE
+	active_power_usage = 2000
+
+/obj/machinery/atmospheric_field_generator/perma/impassable
+	desc = "An older model of ARF-G that generates an impassable retention field whilst powered. This model's field is quite power-efficient, but will still drop from loss of power or electromagnetic interference.<br><br>Note: prolonged immersion in active atmospheric retention fields may have negative long-term health consequences."
+	active_power_usage = 1500
+	field_type = /obj/structure/atmospheric_retention_field/impassable
+
+/obj/machinery/atmospheric_field_generator/attackby(obj/item/W as obj, mob/user as mob)
+	if(W.is_crowbar() && isactive)
+		if(!src) return
+		to_chat(user, "<span class='warning'>You can't open the ARF-G whilst it's running!</span>")
+		return
+	if(W.is_crowbar() && !isactive)
+		if(!src) return
+		to_chat(user, "<span class='notice'>You [hatch_open? "close" : "open"] \the [src]'s access hatch.</span>")
+		hatch_open = !hatch_open
+		update_icon()
+		if(alwaysactive && wires_intact)
+			generate_field()
+		return
+	if(hatch_open && W.is_multitool())
+		if(!src) return
```

Comment:
I covered this in the suggestion above, but if you're doing this at this in every single if statement, you should just do it at the top of the function.